src/utils/middleware/checkHeaders.ts

```diff
@@ -0,0 +1,114 @@
+/**
+ * An express middleware that checks if headers contain some expected values
+ */
+
+/* eslint-disable sonarjs/no-duplicate-string */
+
+import { Request, Response, NextFunction } from "express";
+import { fromNullable, Option } from "fp-ts/lib/Option";
+import { fromPredicate, fromOption, Either } from "fp-ts/lib/Either";
+import { intersection } from "fp-ts/lib/Array";
+import { setoidString } from "fp-ts/lib/Setoid";
+import { log } from "../logger";
+
+type ValidateErrors =
+  | "no-header"
+  | "unexpected-header-values"
+  | "no-expected-headers-provided";
+
+const acceptHeader = (req: Request): Option<string> =>
+  fromNullable(req.headers.accept);
+
+const contentTypeHeader = (req: Request): Option<string> =>
+  fromNullable(req.headers["content-type"]);
+
+export const validate = (expectedHeaderValues: ReadonlyArray<string>) => (
+  maybeHeader: Option<string>
+): Either<ValidateErrors, ReadonlyArray<string>> =>
+  // check if expected headers is not empty
+  fromPredicate<ValidateErrors, void>(
+    () => expectedHeaderValues.length > 0,
+    () => "no-expected-headers-provided"
+  )().chain<ReadonlyArray<string>>(() =>
+    fromOption<ValidateErrors>("no-header")(maybeHeader)
+      // split header on , for values and on ; for q-factors
+      .map(header => header.split(/[,;]/).map(s => s.trim()))
+      // get the common values between header values and expected values
+      .map(headerValues =>
+        intersection(setoidString)(
+          headerValues,
+          // eslint-disable-next-line functional/prefer-readonly-type
+          expectedHeaderValues as string[]
+        )
+      )
+      .chain(
+        fromPredicate(
+          commonValues => commonValues.length > 0,
+          () => "unexpected-header-values"
+        )
+      )
```

Comment:
@balanza there was a .map(_ => void 0) here. I removed it and changed the right side of either to ReadonlyArray<string>.